### PR TITLE
First attempt at animation of features

### LINKED ISSUE #113: [rendering] boss-level feature: migrating geometries
It has always been envisioned (OK, in my head anyway) to have a high-challenge snazzy feature of "migrating geometries". Putting this out there just to get it out of my head and into an issue.

To explain: take the situation of the US State of Virginia. It existed at one point as a merge of current Kentucky, West Virginia, and Virginia. In two successive steps, those states were ceded and Virginia shrunk to its current geometry. This is a non-migrating geometry change. On one day, it contained "three states worth", then the next day, "two states worth", then a subsequent date: one. Ie. very discrete changes.

Now take the situation of the growth of the Roman Empire under Augustus. There are no discrete dates, the empire just grows. (Rightly speaking: given the confidence of the accuracy of records in the 1st century BC, the empire appears to grow amorphously across time as the conquests into the frontiers ebb.) This could be considered a "migrating geometry".

Similarly, literally migratory indigenous tribes have slow-moving amorphous changes to their center of gravity.

While rendering these two types (non-migratory and migratory geometries) one would want a way to visualize these differently. Using the current viewport terminology, the `Advance` button would want to advance a percentage of the way from one geometry to another. This would allow a way to designate a "start of migration" geometry and an "end of migration" geometry, and the rendering tool would manage the morph from the first to the second, based upon where in time the current display is. So, using the Augustus example, there could be a geometry for "Roman Empire in 31BC", then another in 19BC, and the advance button would morph between the two during the time crawl. There would need to be some way in the database to link these two geometries.

It is a "boss-level" feature because it would be complex to render cleanly, especially around coastlines. But might work better for indigenous tribes which currently only designate borderless-areas.

### PR DESCRIPTION
**Pull request by ohmec**
Fixes #113 

Here's a decent rendering of how the animation could work. A few notes in this investigation.

- In the database, a feature is designated to "animateTo" another feature, which has to exist (not enough error checking is added just yet)
- The two features must be identical in coordinate length or the "animation" fails (not enough error checking is added just yet), and the coordinate order must be the same. This wasn't too hard with the KML editor, I just created one, saved it, then moved a subset of its coordinate points, then saved. Unfortunately the KML editor did strange things and didn't save them in the same order, so I hand-adjusted that. But I don't think it is too much to keep this constraint
- I imagine the performance will be a direct function of how many coordinates and features are moving at a time. There is no noticeable lag with just this one feature per time period animating at a time. To reduce compute time it pre-determines which subset of coordinates change during the animation and only alters those.
- It looked a little hokey without also animating the changes between label adjustments, so those are altered to. I.e. if the animateFrom label has rotation of -45, and the animateTo has 30, then throughout the time period of the "from" feature it moves through 75 degrees of rotation.

To see these "in action" I've prepared two demonstration vehicles:

1. label only: Olmec inhabitants from 1500BC to 400BC. To render, add these adjustments to the viewing URL: `?startdatestr=-1550&enddatestr=-300&curdatestr=-1500&lat=18.43&lon=-95.80&z=8.0`
2. polygon and label: Roman incursion into modern Turkey from 44BC to 14AD. To render, add these adjustments to the viewing URL: `?startdatestr=-50&curdatestr=-44&enddatestr=20&lat=39.16&lon=34.66&z=6.0`

Signed-off-by: Scott Johnson <sjcupertino@gmail.com>

### CODE REVIEW

**Comment by warrenjp** [comment]
This is a great first effort. Both examples worked well for me. This is going to be a really important feature for this platform.

Two points:

1) For the Roman Incursion example, how does the animation work if you have two adjacent regions dynamically changing over time? Does the movement of both regions look good? What about three or more adjacent features? 

2) We know that the final state of the animation is influencing the shape of the intermediate region boundaries. This creates some odd intermediate boundary shapes that don't make sense (although, they're not that bad). For example, if you move to 02/08/0010 in the Roman Incursion, you will see a point that sticks out into the Black Sea at the northeast point of the incursion. This is an artifact of the final destination of the incursion which will end up swinging up the coast into Georgia by 01/01/0014. 

A more "accurate" animation wouldn't extend out into the Black Sea as it dynamically moves over to its final destination, but would rather adhere to the coast and then finally move up the coast and around into Georgia as it nears the final date. To do this, we could take into account the location of all the points along all the coasts and insure that the animation never breaks through these coastal boundaries. This is no simple task. All the same, we would only have to identify these points once. Another option is to take into consideration the final boundaries of the region and insure that the intermediate boundaries never break outside of these final boundaries. Since the final boundaries have an understanding of the coastal borders, this would insure that the intermediate boundaries never "leak out" beyond the coasts.

Either way, these strategies would take additional computing power and perhaps impact performance to too great of an extent. Nevertheless, it's probably worthwhile to test these options out to see how they perform.

**Comment by ohmec** [comment]
@warrenjp I've added a more complex set of animations here to show how they can work together. For best demonstration, add these modifiers:

`?startdatestr=-15000:01:01&enddatestr=-6000:12:31&curdatestr=-15000:01:01&lat=58.82&lon=2.85&z=4.0`

A couple of notes: a) the data is roughly accurate for the Magdalenian and Ahrsburg cultures, plus the icecaps. The Maglemosian is more a combination of guesswork and showing how to work "around corners". b) this introduces a new concept called a "geographical entity", in this case the icecaps. Related: c) the coastlines were quite different in the ice age due to much lower sea levels. To get this time period right we'd probably have to have a geographical entity for modified coastlines. Of course the "Phase N" solution would have these included beautifully rendered in the underlying tile.

**Comment by ohmec** [comment]
Forgot to mention d) yes these features shouldn't be in `ohmec_data_na.js`. I'll spin them out to `ohmec_data_eur.js` and leave them shelved for a future time, after completing NA and Phase I.

**Review by warrenjp — APPROVED** [review]
Looks really cool! Worked well on my machine, so I'm approving. I did not get into the details of the implementation, so I can't comment on the approach that you took in the code.

**Comment by ohmec** [comment]
OK, converting this from draft to real now that I've moved the prehistoric stuff out of `ohmec_data_na.js` (and into `ohmec_data_eur.js` which is just as if sitting on a shelf for now.) I'll update the documentation to describe `animateTo` now, then merge.

File: geo_features.js [hunk 1]
```diff
@@ -142,6 +143,32 @@ function featureStyle(feature) {
     strokeWeight = 1.5;
     strokeDash   = '1';
     fillColor    = '#031056';
+  } else if(entity1name == 'Roman Empire') {
```
Comment (ohmec) [inline geo_features.js:146]: ooh, now that I'm seeing these here, I'm thinking the corresponding `ohmec_data_<>.js` file should be the one to define these. that way they're in the author's control, not a random disassociated file. Will file an issue but leave for now.

**Comment by ohmec** [comment]
The documentation has been updated. I will merge this tomorrow.

**Pull request merged**